### PR TITLE
feat: `std::hint::black_box` function.

### PR DESCRIPTION
**Pull request by aakoshh**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/6471

## Summary\*

Added a `std::hint` Noir module with a `black_box` builtin function that is treated during SSA optimisation steps as if it could potentially have side effects, which prevents the value it is applied to from being simplified to a known value and the instruction from being removed. 

## Additional Context

Using a `builtin` rather than `foreign` so that this can be called from non-Brillig code. 

I added an `Intrinsic::Hint` because `Intrinsic::BlackBox` was already taken; perhaps `Intrinsic::AsWitness` could be moved into `Hint` as well (not that it matters, but the comments call it a hint too).

## Documentation\*

Check one:
- [ ] No documentation needed.
- [x] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: test_programs/execution_success/hint_black_box/src/main.nr
Comment (TomAFrench) [inline test_programs/execution_success/hint_black_box/src/main.nr:1]: Can you add some tests to check how it handles arrays, etc.?

Comment (aakoshh) [reply test_programs/execution_success/hint_black_box/src/main.nr:1]: Good shout, the slice tests uncovered a bug.

Comment (aakoshh) [reply test_programs/execution_success/hint_black_box/src/main.nr:1]: Added some tests with mutable references but that doesn't work very well. I left some comments about scenarios that didn't work out, let me know if you think there's anything we can do about them.

**Comment by TomAFrench** [comment]
I'd like to take another look at this today after scrum

**Comment by aakoshh** [comment]
@TomAFrench have you had a chance to look at this?

**Review by TomAFrench — APPROVED** [review]

**Comment by TomAFrench** [comment]
Thanks, sorry for the delay on this one.

**Comment by Savio-Sou** [comment]
Looks like this is not currently documented in the Noir docs.

Is the new function potentially useful for the community?

**Pull request merged**